File: captainhcg/js/410-split-array-largest-sum.js
var splitArray = function(nums, m) {
    if(nums.length === 1){
        return nums[0];
    }
    var min = Math.max.apply(null, nums);
    var max = nums.reduce(function(a, b){ return a+b}, 0)
    var l = min, r = max, valid = max;
    var checkValid = function(sum){
        var _sum = 0, _m = 1;
        for(var n of nums){
            if(_sum + n <= sum){
                _sum += n;
            } else {
                _m += 1
                if(_m > m){
                    return false
                }
                _sum = n;
            }
        }
        return true;
    }
    while(l <= r){
        var mid = Math.ceil((l + r) / 2);
        if(checkValid(mid)){
            r = mid - 1;
            valid = mid;
        } else {
            l = mid + 1;
        }
    }
    return valid;
};

File: bwu/binary_search/410-split-array-largest-sum.py
class Solution(object):
    def splitArray(self, nums, m):
        """
        :type nums: List[int]
        :type m: int
        :rtype: int
        """
        def isValid(val):
            cnt, tmp = 0, 0
            for n in nums:
                tmp += n
                if tmp > val:
                    cnt +=1
                    if cnt >= m:
                        return False
                    tmp = n
            return True
            
        left, right = max(nums), sum(nums)
        while left + 1 < right:
            mid = (left + right) >> 1
            if isValid(mid):
                right = mid
            else:
                left = mid
                
        if isValid(left):
            return left
        else:
            return right
